### PR TITLE
KIALI-945 Swagger documentation

### PR DESCRIPTION
**Pull request by aljesusg**
swagger

### CODE REVIEW

**Comment by lucasponce** [comment]
I would like to play a little bit more with this.
In Hawkular Alerting we introduced our scripts and annotations
https://github.com/hawkular/hawkular-alerts/tree/master/rest-handlers/src/main/rest-doc
Basically due the lack of flexibility of swagger to deal with Java -> JSON.

But this looks ok, it seems as the json is already serialized that would be simplified.

**Comment by aljesusg** [comment]
Mmm I don't have experience with Java but I'm curious about these problems, with go it's easy just comment and use go-swagger ^^ if you need help ping me @lucasponce

**Comment by aljesusg** [comment]
Arhh don't merge this swagger.json should be in the .gitignore, we should generate swagger.json with Travis and upload to another place.

**Comment by lucasponce** [comment]
You can take a look at the result
http://www.hawkular.org/docs/rest/rest-alerts-v2.html
Basically swagged can't deal well mapping correctly types, allowable values, notes.
We had too many exceptions and custom mappings that made swagger (the java lib) less useful for our work. 

That was particularly useful in the complex types:
http://www.hawkular.org/docs/rest/rest-alerts-v2.html#Trigger

Which are json, but we needed this extra layer of field information, as at the end it was not easy to add these notes.

**Comment by lucasponce** [comment]
Adding @JSticker as I think she is in contact with other groups and probably she can give some advice in the how to deal with the generated doc.

File: Makefile [hunk 1]
```diff
@@ -144,6 +144,24 @@ dep-update:
...
+# Swagger Documentation
+#
+swagger-install:
+	go get -u github.com/go-swagger/go-swagger/cmd/swagger
+
+swagger-validate:
+	@swagger validate ./swagger.json
+	@if [ "$$?" != "0" ]; then \
+        exit 1; \
+    fi
```
Comment (simonpasquier) [inline Makefile]: nit but the `if ...` part is unnecessary since `make` already returns the exit code in case the swagger command fails.

**Comment by aljesusg** [comment]
I found the [swagger-codegen ](https://github.com/swagger-api/swagger-codegen) this allow  us generate de website like static index.html and it seems that we can customize the template. What do you think @JSticker ?

html

File: config/token.go [hunk 2]
```diff
@@ -18,8 +18,36 @@ type TokenClaim struct {
 	jwt.StandardClaims
 }
 
+// HTTP status code 200 and tokenGenerated model in data
+// swagger:response tokenGenerated
+type swaggTokenGeneratedResp struct {
```
Comment (lucasponce) [inline config/token.go]: I am starting to review. Do handlers need a refactor to add a wrapper for the response ?

File: kiali.go [hunk 3]
```diff
@@ -1,3 +1,18 @@
+// Kiali
+//
+// Kiali project, observability for the Istio service mesh
+//
+//     Schemes: http
```
Comment (lucasponce) [inline kiali.go]: Minor: "Or https, as by default OS traffic is secured."

File: kiali.go [hunk 4]
```diff
@@ -1,3 +1,18 @@
...
+//     BasePath: /api
+//     Version: 0.2.0.Alpha
```
Comment (lucasponce) [inline kiali.go]: We will need a good way to update this on releases.
It will be pretty easy to not put the correct version here.
Perhaps this needs a comment in the README or in the releases doc.

Comment (lucasponce) [reply kiali.go]: cc @jmazzitelli

File: routing/routes.go [hunk 5]
```diff
@@ -25,27 +25,85 @@ func NewRoutes() (r *Routes) {
 	r = new(Routes)
 
 	r.Routes = []Route{
+		// swagger:route GET / Root
+		// ---
+		// Handler to get the status of Kiali
```
Comment (lucasponce) [inline routing/routes.go]: If this is going to the documentation, I would use "Endpoint" rather than "Handler".
As end user Endpoint is clear, "handler" is more the terminology used to implement it (I think).

Comment (lucasponce) [reply routing/routes.go]: "Handler" is used in several locations, I won't mark them all of them to avoid noise.

File: routing/routes.go [hunk 6]
```diff
@@ -25,27 +25,85 @@ func NewRoutes() (r *Routes) {
...
+		//
+		//     Consumes:
+		//     - application/json
+		//
+		//     Produces:
+		//     - application/json
+		//
+		//     Schemes: http
+		// responses:
+		//      200: statusInfo
```
Comment (lucasponce) [inline routing/routes.go]: I guess this "statusInfo" will be resolved by Swagger in the specific type. Isn't it ?

File: routing/routes.go [hunk 7]
```diff
@@ -25,27 +25,85 @@ func NewRoutes() (r *Routes) {
...
+		{
 			"Status",
 			"GET",
 			"/api/status",
 			handlers.Root,
 			false,
 		},
+		// swagger:route GET /namespaces/{namespace}/istio istioConfigList
```
Comment (lucasponce) [inline routing/routes.go]: The param "namespace" needs to be explained.

File: routing/routes.go [hunk 8]
```diff
@@ -25,27 +25,85 @@ func NewRoutes() (r *Routes) {
...
+		//     Consumes:
+		//     - application/json
+		//
+		//     Produces:
+		//     - application/json
+		//
+		//     Schemes: http
+		//
+		//     responses:
+		//       200: istioConfigList
```
Comment (lucasponce) [inline routing/routes.go]: Also, good to add the generic error responses.
On this case a 404 is possible and in generic, a 500 should be also a valid error in all endpoints.
For example, in Hawkular we covered all
http://www.hawkular.org/docs/rest/rest-alerts-v2.html#__alerts_handling

Comment (lucasponce) [reply routing/routes.go]: This comment may apply to the other endpoints as well.

File: config/token.go (hunk 2 above)
Comment (aljesusg) [reply config/token.go]: No, It's an addition for swagger documentation, No refactor need it

File: kiali.go (hunk 3 above)
Comment (aljesusg) [reply kiali.go]: Actually the API not works with https only http

File: services/models/destination_policy.go [hunk 9]
```diff
@@ -4,10 +4,32 @@ import (
 	"github.com/kiali/kiali/kubernetes"
 )
 
+// DestinationPolicies destinationPolicies
+//
+// This is used for returning an array of DestinationPolicies
```
Comment (lucasponce) [inline services/models/destination_policy.go]: Minor suggestion about style.
For documentation I find useful to be explicit and about pronoums.
Instead "This is used" ... > "This type/endpoint/function/... is used for" 
That would help to avoid ambiguity.
It is more verbose but at the end it helps.
(Same for other areas)

File: kiali.go (hunk 4 above)
Comment (aljesusg) [reply kiali.go]: +1 This version only set this information in https://kiali.org/pages/api/overview.html 

Version information

Version : 0.2.0.Alpha

File: services/models/destination_policy.go [hunk 10]
```diff
@@ -4,10 +4,32 @@ import (
...
+//
+// swagger:model destinationPolicies
+// An array of destinationPolicie
```
Comment (lucasponce) [inline services/models/destination_policy.go]: Typo: destinationPolicie -> destinationPolicy

File: routing/routes.go (hunk 6 above)
Comment (aljesusg) [reply routing/routes.go]: Yes is like a tag this go to the swagger: response swagger:model called statusInfo

File: routing/routes.go (hunk 8 above)
Comment (aljesusg) [reply routing/routes.go]: OK I'll work on the genericResponse, I'll add a new struct for this like "Response", No refactor need it

File: services/models/destination_policy.go [hunk 11]
```diff
@@ -4,10 +4,32 @@ import (
...
+//
+// swagger:model destinationPolicy
 type DestinationPolicy struct {
-	Name            string      `json:"name"`
-	CreatedAt       string      `json:"createdAt"`
+	// The name of the destinationPolicy
+	//
+	// required: true
+	Name string `json:"name"`
+	// The creation date of the destinationPolicy
```
Comment (lucasponce) [inline services/models/destination_policy.go]: Good if fields with format expected can define a note about this.
"createdAt should be a string with format XYZ"
Also, when field is a string, but it is expected a known value, potential values should be described.

This comment is not to fill all the fields, but to enable swagger to add these kind of info (we can fix them on future iterations).

For example the Severity field on this one it can accept "LOW, MEDIUM, HIGH, CRITICAL":
http://www.hawkular.org/docs/rest/rest-alerts-v2.html#Trigger

A potential 400 error can be send back if there is some incorrect value.

Basically we need to document these cases as well.

File: services/models/destination_rule.go [hunk 12]
```diff
@@ -26,7 +48,7 @@ func (dRule *DestinationRule) Parse(destinationRule kubernetes.IstioObject) {
 	dRule.Name = destinationRule.GetObjectMeta().Name
 	dRule.CreatedAt = formatTime(destinationRule.GetObjectMeta().CreationTimestamp.Time)
 	dRule.ResourceVersion = destinationRule.GetObjectMeta().ResourceVersion
-	dRule.DestinationName = destinationRule.GetSpec()["name"]
+	dRule.Host = destinationRule.GetSpec()["host"]
```
Comment (lucasponce) [inline services/models/destination_rule.go]: Mmm, did you rebase with master ?
In theory, these changes should be taken from the master, not fixed on these PR unless there is a bug.
Perhaps you would get some conflict then.

File: services/models/namespace.go [hunk 13]
```diff
@@ -6,7 +6,27 @@ import (
 	"github.com/kiali/kiali/kubernetes"
 )
 
+// A Namespace is the main classification in kiali.
```
Comment (lucasponce) [inline services/models/namespace.go]: I guess we should align the terms with kubernetes, istio and/or openshift.
As we use the "Namespace" concept in the same way as k8s, nothing wrong to re-use same wording from their API for same things (i.e. namespace).
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#namespace-v1-core

File: services/models/route_rule.go [hunk 14]
```diff
@@ -4,10 +4,33 @@ import (
...
+// This is used for returning an array of RouteRule
+//
+// swagger:model routeRules
+// An array of routeRule
+// swagger:allOf
 type RouteRules []RouteRule
+
+// RouteRule routeRule
+//
+// This is used for returning a RouteRule
```
Comment (lucasponce) [inline services/models/route_rule.go]: On these cases, where we are also describing practically an object from Istio, we can use a reference to their doc, too.
I mean, swagger should be able to enable these notes.
Developers can maintain them in an iterative way, but that is something we want to set up.

File: swagger.json [hunk 15]
```diff
@@ -0,0 +1,606 @@
+{
```
Comment (lucasponce) [inline swagger.json]: In Hawkular the swagger.json was generated but not committed in the PR.
Travis was the responsible to generate it, transform it, and commit it in the repo
https://github.com/hawkular/hawkular.github.io/tree/swagger

I have no preference, but just wonder if we can re-use something from hawkular repo, or if it is simple to commit the swagger.json with the code.

File: services/models/destination_policy.go (hunk 11 above)
Comment (aljesusg) [reply services/models/destination_policy.go]: Yes we can set // pattern: \w[\w-]+

File: services/models/destination_rule.go (hunk 12 above)
Comment (aljesusg) [reply services/models/destination_rule.go]: Maybe was the change from istio-08 to master :'(

**Comment by aljesusg** [comment]
The make task is ready to get the Json from Kiali master 

```bash
swagger-doc:
	@java -jar ./scripts/swagger2markup-cli-1.3.3.jar convert \
	 -i https://raw.githubusercontent.com/kiali/kiali/master/swagger.json \
	 -d ./pages/api/ \
	 -c ./scripts/config.properties
```

I'll improve the documentation in another PR and I'll do the Travis job after Hugo release.

**Comment by theute** [comment]
I was talking about this one: https://github.com/kiali/kiali.io/blob/master/scripts/build-swagger.sh
We can update it  once this PR is merged.

The script is ran on all merge (or Travis rebuild) in kiali.io repo

**Comment by aljesusg** [comment]
Oo I didn't see it the kiali.io repo ^^ Great if Travis is ready @theute, so we can merge

**Comment by JStickler** [comment]
Hey guys, sorry I didn't respond to the mention on this, I was on PTO and offline.

File: config/token.go (hunk 2 above)
Comment (xeviknal) [reply config/token.go]: pretty late here, but do you think is a way to avoid the fact of having to add a new type like this one for each API endpoint?

Comment (xeviknal) [reply config/token.go]: yep, those definitions are important, sure thing :) My point is that we only have one piece of software that handles all the responses in an generic way. If there is no other option, perhaps we should create a folder where we place all this only-for-documentation code.
Let's see what they tell you :)

Comment (aljesusg) [reply config/token.go]: Good idea +1
